compositor_pipeline/src/pipeline/input/rtp/depayloader.rs

```diff
@@ -204,9 +218,273 @@ impl RolloverState {
+/// MPEG-4 part 3, 4.5.1.1
+fn frame_length_flag_to_frame_length(flag: bool) -> u32 {
+    match flag {
+        true => 960,
+        false => 1024,
+    }
+}
+
+struct Asc {
+    _profile: u8,
+    frequency: u32,
+    _channel: u8,
+    frame_length: u32,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum AudioSpecificConfigParseError {
+    #[error("ASC is not long enough")]
+    TooShort,
+
+    #[error("Illegal value in ASC")]
+    IllegalValue,
+}
+
+// MPEG-4 part 3, sections 1.6.2.1 & 4.4.1
+fn parse_asc(asc: &[u8]) -> Result<Asc, AudioSpecificConfigParseError> {
+    // TODO: this can probably be rewritten using [nom](https://lib.rs/crates/nom), which would
+    // make it a lot more understandable
+    let mut reader = std::io::Cursor::new(asc);
+
+    if reader.remaining() < 2 {
+        return Err(AudioSpecificConfigParseError::TooShort);
+    }
+
+    let first = reader.get_u8();
+    let second = reader.get_u8();
+
+    let mut profile = (0b11111000 & first) >> 3;
+    let frequency: u32;
+    let channel: u8;
+    let frame_length: u32;
+
+    if profile == 31 {
+        profile = ((first & 0b00000111) << 3) + ((second & 0b11100000) >> 5) + 32;
+        let frequency_id = (second & 0b00011110) >> 1;
+
+        let channel_and_frame_len_bytes: [u8; 2];
+
+        if frequency_id == 15 {
+            if reader.remaining() < 4 {
+                return Err(AudioSpecificConfigParseError::TooShort);
+            }
+
+            let mut rest = [0; 4];
+            reader.read_exact(&mut rest).unwrap();
+
+            frequency = (((second & 0b00000001) as u32) << 23)
+                | ((rest[0] as u32) << 15)
+                | ((rest[1] as u32) << 7)
+                | (((rest[2] & 0b11111110) >> 1) as u32);
+
+            channel_and_frame_len_bytes = [rest[2], rest[3]];
+        } else {
+            if reader.remaining() < 1 {
+                return Err(AudioSpecificConfigParseError::TooShort);
+            }
+            let last = reader.get_u8();
+
+            channel_and_frame_len_bytes = [second, last];
+            frequency = freq_id_to_freq(frequency_id)?
+        };
+
+        let [b1, b2] = channel_and_frame_len_bytes;
+        channel = ((b1 & 0b00000001) << 3) | ((b2 & 0b11100000) >> 5);
+        let frame_length_flag = b2 & 0b00010000 != 0;
+
+        frame_length = frame_length_flag_to_frame_length(frame_length_flag);
+    } else {
+        let frequency_id = ((first & 0b00000111) << 1) + ((second & 0b10000000) >> 7);
+        let channel_and_frame_len_byte: u8;
+
+        if frequency_id == 15 {
+            if reader.remaining() < 3 {
+                return Err(AudioSpecificConfigParseError::TooShort);
+            }
+
+            let mut rest = [0; 3];
+            reader.read_exact(&mut rest).unwrap();
+            frequency = (((second & 0b01111111) as u32) << 17)
+                | ((rest[0] as u32) << 9)
+                | ((rest[1] as u32) << 1)
+                | (((rest[2] & 0b10000000) >> 7) as u32);
```

Comment:
I would order those operation based on what bits they are setting e.g.
```suggestion
            frequency = (((rest[2] & 0b10000000) >> 7) as u32)
                | ((rest[1] as u32) << 1)
                | ((rest[0] as u32) << 9)
                | (((second & 0b01111111) as u32) << 17);
```